File: frontend/src/api/secureStore.ts
// Secure storage for the one real secret the client holds: the bearer token.
// - Native: expo-secure-store → iOS Keychain / Android Keystore-backed store (encrypted at rest,
//   not readable from a plain ADB backup or by other apps).
// - Web (localhost preview only): falls back to localStorage (no Keychain in a browser).
// A memory fallback keeps auth working if the secure module ever throws.
import { Platform } from "react-native";

const mem = new Map<string, string>();

let SecureStore: typeof import("expo-secure-store") | null = null;
if (Platform.OS !== "web") {
  try {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    SecureStore = require("expo-secure-store");
  } catch {
    SecureStore = null;
  }
}

const webLS = (): Storage | null => {
  try {
    if (typeof globalThis !== "undefined" && (globalThis as any).localStorage) {
      return (globalThis as any).localStorage as Storage;
    }
  } catch {}
  return null;
};

export const secureStore = {
  async get(key: string): Promise<string | null> {
    try {
      if (SecureStore) {
        const v = await SecureStore.getItemAsync(key);
        if (v != null) return v;
      } else {
        const ls = webLS();
        if (ls) {
          const v = ls.getItem(key);
          if (v != null) return v;
        }
      }
    } catch {}
    return mem.has(key) ? (mem.get(key) as string) : null;
  },
  async set(key: string, value: string): Promise<void> {
    mem.set(key, value);
    try {
      if (SecureStore) await SecureStore.setItemAsync(key, value);
      else webLS()?.setItem(key, value);
    } catch {}
  },
  async remove(key: string): Promise<void> {
    mem.delete(key);
    try {
      if (SecureStore) await SecureStore.deleteItemAsync(key);
      else webLS()?.removeItem(key);
    } catch {}
  },
};

File: flutter_app/lib/services/secure_store.dart
import 'package:flutter/foundation.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';

/// Encrypted storage for the one real secret this client holds: the bearer token.
///
/// The session used to be persisted whole — token included — as a JSON blob in
/// SharedPreferences. That file is plaintext on disk: readable from an ADB backup, from
/// another app on a rooted device, or by anyone with the handset. The token now lives here
/// instead (iOS Keychain / Android EncryptedSharedPreferences), and the profile blob in
/// SharedPreferences is written without it.
///
/// Web has no OS keychain, so `flutter_secure_storage` falls back to browser storage there.
/// That is unavoidable in a browser and matches how the Expo app behaves; on Android and
/// iOS — where the APK actually ships — the token is encrypted at rest.
class SecureStore {
  static const _storage = FlutterSecureStorage(
    aOptions: AndroidOptions(encryptedSharedPreferences: true),
    iOptions: IOSOptions(accessibility: KeychainAccessibility.first_unlock),
  );

  /// In-memory fallback so auth still works if the platform channel throws
  /// (older Android keystores occasionally do). Never written to disk.
  static final Map<String, String> _memory = {};

  static Future<String?> read(String key) async {
    try {
      final v = await _storage.read(key: key);
      if (v != null) return v;
    } catch (e) {
      debugPrint('SecureStore.read failed: $e');
    }
    return _memory[key];
  }

  static Future<void> write(String key, String value) async {
    _memory[key] = value;
    try {
      await _storage.write(key: key, value: value);
    } catch (e) {
      debugPrint('SecureStore.write failed: $e');
    }
  }

  static Future<void> delete(String key) async {
    _memory.remove(key);
    try {
      await _storage.delete(key: key);
    } catch (e) {
      debugPrint('SecureStore.delete failed: $e');
    }
  }
}
